components/org.wso2.carbon.metrics.impl/src/main/java/org/wso2/carbon/metrics/impl/MetricServiceImpl.java

```diff
@@ -378,10 +417,112 @@ private boolean isMetricEnabledBasedOnHierarchyLevel(String name, Level metricLe
+            List<?> affected = getAffectedMetrics(annotatedName, levels, metricBuilder);
+            if (annotated && metric instanceof Counter) {
+                metricCollection = new CounterCollection((Counter) metric, (List<Counter>) affected);
+                metricsCollections.put(annotatedName, metricCollection);
+            } else if (annotated && metric instanceof Histogram) {
+                metricCollection = new HistogramCollection((Histogram) metric, (List<Histogram>) affected);
+                metricsCollections.put(annotatedName, metricCollection);
+            } else if (annotated && metric instanceof Meter) {
+                metricCollection = new MeterCollection((Meter) metric, (List<Meter>) affected);
+                metricsCollections.put(annotatedName, metricCollection);
+            } else {
+                metricCollection = metric;
+            }
+        }
+        return metricCollection;
+    }
+
+    private boolean isAnnotated(String annotatedName) {
+        return annotatedName.contains("[+]");
+    }
+
+    private boolean isLevelsMatch(String annotatedName, Level[] levels) {
+        Pattern p = Pattern.compile("\\[\\+\\]");
+        Matcher m = p.matcher(annotatedName);
+        int affectedMetrics = 0;
+        while (m.find()) {
+            affectedMetrics++;
+        }
+        // Levels count should be equals to affected metrics + current metric
+        return levels.length == affectedMetrics + 1;
+    }
+
+    /**
+     * Get affected Metrics for a given hierarchy path
+     *
+     * @param annotatedName The annotated name of the metric
+     * @param levels The {@code Level}s for affected metrics
+     * @param metricBuilder A {@code MetricBuilder} instance used to create the relevant metric
+     * @return The created {@link List<Metric>} collection
+     */
+    private <T extends AbstractMetric> List<?> getAffectedMetrics(String annotatedName, Level[] levels,
+            MetricBuilder<T> metricBuilder) throws MetricNotFoundException {
+        boolean getOrCreate = (levels != null) && (levels.length > 0);
+        int levelIndex = 0;
+        int index = annotatedName.lastIndexOf(".");
+        String annotatedPath = annotatedName.substring(0, index);
+        String statName = annotatedName.substring(index + 1);
+
+        List<T> affected = new ArrayList<>();
+        String[] chunks = annotatedPath.split("\\.");
+        StringBuilder builder = new StringBuilder();
+        String affectedName;
+        // i < chunksLength - 1, cause affected metrics for path org.wso2.product.stat-name
+        // will be searched only in org.stat-name, org.wso2.stat-name.
+        for (int i = 0, chunksLength = chunks.length; i < chunksLength - 1; i++) {
+            String chunk = chunks[i];
+            if (builder.length() > 0) {
+                builder.append('.');
+            }
+            builder.append(chunk);
+            if (chunk.contains("[+]")) {
```

Comment:
@grainier Can you name the string constants appropriately and keep them in a separate constant class? Since these are symbols denote meaning within the context of the component and are used everywhere, better define them at one place and reuse that everywhere.